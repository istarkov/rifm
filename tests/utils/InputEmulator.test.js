/* @flow */

import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import { Value } from 'react-powerplug';
import { InputEmulator, renderInputState } from './InputEmulator';

declare var test: Function;
declare var expect: Function;
declare var jest: Object;

test('Input emulator commands test', () => {
  let getVal = null;
  let execCommand = null;
  let snaphot = [];
  let reactVal = '';

  TestRenderer.create(
    <Value
      initial=""
      onChange={v => {
        reactVal = v;
      }}
    >
      {input => (
        <InputEmulator
          value={input.value}
          onChange={event => input.set(event.target.value)}
        >
          {(exec, val) => {
            execCommand = exec;
            getVal = val;
            return null;
          }}
        </InputEmulator>
      )}
    </Value>
  );

  const exec = cmd => {
    if (!execCommand || !getVal) {
      throw Error('rifm is not initialized');
    }

    execCommand(cmd);
    snaphot.push({ ...getVal(), cmd, withCaret: renderInputState(getVal()) });
  };

  exec({ type: 'PUT_SYMBOL', payload: '1' });
  exec({ type: 'PUT_SYMBOL', payload: '34' });
  exec({ type: 'MOVE_CARET', payload: -2 });
  exec({ type: 'PUT_SYMBOL', payload: '2' });
  exec({ type: 'MOVE_CARET', payload: 100 });
  exec({ type: 'PUT_SYMBOL', payload: '5' });
  exec({ type: 'MOVE_CARET', payload: -100 });
  exec({ type: 'PUT_SYMBOL', payload: '0' });
  exec({ type: 'PUT_SYMBOL', payload: 'd' });
  exec({ type: 'BACKSPACE' });
  exec({ type: 'BACKSPACE' });
  exec({ type: 'BACKSPACE' });
  exec({ type: 'MOVE_CARET', payload: 100 });
  exec({ type: 'BACKSPACE' });
  exec({ type: 'MOVE_CARET', payload: -1 });
  exec({ type: 'BACKSPACE' });
  exec({ type: 'PUT_SYMBOL', payload: '3' });
  exec({ type: 'MOVE_CARET', payload: 100 });
  exec({ type: 'DELETE' });
  exec({ type: 'MOVE_CARET', payload: -1 });
  exec({ type: 'DELETE' });
  exec({ type: 'MOVE_CARET', payload: -2 });
  exec({ type: 'DELETE' });
  exec({ type: 'MOVE_CARET', payload: -100 });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });

  expect(reactVal).toEqual(getVal && getVal().value);

  expect(snaphot).toMatchSnapshot();
});

test('Input emulator work as React if values dont match', () => {
  let getVal = null;
  let execCommand = null;
  let snaphot = [];
  let badSymbol = '';

  TestRenderer.create(
    <Value initial="">
      {input => (
        <InputEmulator
          value={badSymbol + input.value}
          onChange={event => input.set(event.target.value)}
        >
          {(exec, val) => {
            execCommand = exec;
            getVal = val;
            return null;
          }}
        </InputEmulator>
      )}
    </Value>
  );

  const exec = cmd => {
    if (!execCommand || !getVal) {
      throw Error('rifm is not initialized');
    }

    execCommand(cmd);
    snaphot.push({ ...getVal(), cmd, withCaret: renderInputState(getVal()) });
  };

  exec({ type: 'PUT_SYMBOL', payload: '1' });
  exec({ type: 'PUT_SYMBOL', payload: '2' });
  exec({ type: 'MOVE_CARET', payload: -100 });
  badSymbol = '-';
  // cursor should move at the end of input
  exec({ type: 'PUT_SYMBOL', payload: '3' });

  expect(snaphot).toMatchSnapshot();
});

test('renderInputState test', () => {
  const snaphot = [];

  snaphot.push(
    renderInputState({
      value: 'hello',
      selectionStart: 1,
      selectionEnd: 4,
    })
  );

  snaphot.push(
    renderInputState({
      value: 'hello',
      selectionStart: 0,
      selectionEnd: 5,
    })
  );

  snaphot.push(
    renderInputState({
      value: 'hello',
      selectionStart: 5,
      selectionEnd: 5,
    })
  );

  snaphot.push(
    renderInputState({
      value: 'hello',
      selectionStart: 0,
      selectionEnd: 0,
    })
  );

  snaphot.push(
    renderInputState({
      value: 'hello',
      selectionStart: 2,
      selectionEnd: 2,
    })
  );

  expect(() =>
    renderInputState({
      value: 'hello',
      selectionStart: 2,
      selectionEnd: 1,
    })
  ).toThrow();

  expect(snaphot).toMatchSnapshot();
});
