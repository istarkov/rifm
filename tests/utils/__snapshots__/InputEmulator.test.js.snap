// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input emulator commands test 1`] = `
Array [
  Object {
    "cmd": Object {
      "payload": "1",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "1",
    "withCaret": "1|",
  },
  Object {
    "cmd": Object {
      "payload": "34",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "134",
    "withCaret": "134|",
  },
  Object {
    "cmd": Object {
      "payload": -2,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "134",
    "withCaret": "1|34",
  },
  Object {
    "cmd": Object {
      "payload": "2",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": "1234",
    "withCaret": "12|34",
  },
  Object {
    "cmd": Object {
      "payload": 100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "1234",
    "withCaret": "1234|",
  },
  Object {
    "cmd": Object {
      "payload": "5",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 5,
    "selectionStart": 5,
    "value": "12345",
    "withCaret": "12345|",
  },
  Object {
    "cmd": Object {
      "payload": -100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "12345",
    "withCaret": "|12345",
  },
  Object {
    "cmd": Object {
      "payload": "0",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "012345",
    "withCaret": "0|12345",
  },
  Object {
    "cmd": Object {
      "payload": "d",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": "0d12345",
    "withCaret": "0d|12345",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "012345",
    "withCaret": "0|12345",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "12345",
    "withCaret": "|12345",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "12345",
    "withCaret": "|12345",
  },
  Object {
    "cmd": Object {
      "payload": 100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 5,
    "selectionStart": 5,
    "value": "12345",
    "withCaret": "12345|",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "1234",
    "withCaret": "1234|",
  },
  Object {
    "cmd": Object {
      "payload": -1,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "1234",
    "withCaret": "123|4",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": "124",
    "withCaret": "12|4",
  },
  Object {
    "cmd": Object {
      "payload": "3",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "1234",
    "withCaret": "123|4",
  },
  Object {
    "cmd": Object {
      "payload": 100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "1234",
    "withCaret": "1234|",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "1234",
    "withCaret": "1234|",
  },
  Object {
    "cmd": Object {
      "payload": -1,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "1234",
    "withCaret": "123|4",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "123",
    "withCaret": "123|",
  },
  Object {
    "cmd": Object {
      "payload": -2,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "123",
    "withCaret": "1|23",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "13",
    "withCaret": "1|3",
  },
  Object {
    "cmd": Object {
      "payload": -100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "13",
    "withCaret": "|13",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "3",
    "withCaret": "|3",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "",
    "withCaret": "|",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "",
    "withCaret": "|",
  },
]
`;

exports[`Input emulator work as React if values dont match 1`] = `
Array [
  Object {
    "cmd": Object {
      "payload": "1",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "1",
    "withCaret": "1|",
  },
  Object {
    "cmd": Object {
      "payload": "2",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": "12",
    "withCaret": "12|",
  },
  Object {
    "cmd": Object {
      "payload": -100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "12",
    "withCaret": "|12",
  },
  Object {
    "cmd": Object {
      "payload": "3",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "-312",
    "withCaret": "-312|",
  },
]
`;

exports[`renderInputState test 1`] = `
Array [
  "h[ell]o",
  "[hello]",
  "hello|",
  "|hello",
  "he|llo",
]
`;
